### ISSUE TITLE
Trying to understand MergeKeys() behavior

### ISSUE THREAD

**Issue by Isopod** [xkbcommon/libxkbcommon#209]
I'm trying to understand how key definitions are merged, but looking at the source code for [`MergeKeys()`](https://github.com/xkbcommon/libxkbcommon/blob/069777f47fbb0393f88e114f101be60504dd367c/src/xkbcomp/symbols.c#L357), I am left a little confused.

Here is my current understanding: Each `KeyInfo` has a merge mode, which can be any of {`override`, `replace`, `augment`}. Let's say I want to merge `KeyInfo` `b` into `KeyInfo` `a`. 
* If `b`'s merge mode is `override`, then all properties carrying a value in `b` will overwrite the respective properties in `a`.
* If `b`'s merge mode is `augment`, only properties will be added to `a` that don't already have a value in `a`. 
* If `b`'s merge mode is `replace`, then all properties are first deleted from `a` and subsequently replaced by `b`'s properties.

So far, so good. My question is: What happens to the merge mode itself? If I merge `b` into `a`, does the result inherit `a`'s merge mode or `b`'s merge mode?

To make it more concrete (*):

```
xkb_symbols "foo" {
  replace key <ABCD> {['a', 'A']}
}
xkb_symbols "bar" {
  key <ABCD> {'b', 'B']}
  include "foo"
}
xkb_symbols "baz" {
  key <ABCD> {[], [['c', 'C']}
  include "bar"
}
```

Does `bar`'s definition inherit the `replace` mode from `foo` or not? Should `baz`'s key definition then be equivalent to 
* ```key <ABCD> {['a', 'A']}``` or 
* ```key <ABCD> {['a', 'A'], ['c', 'C']}``` ?

What about other mode combinations? For example, what if `foo`'s mode was `augment` instead of `replace`? And so on.

Reading the source code, it looks to me like the current behavior is: When merging `b` into `a`, the result inherits
* `b`'s mode if `b` has the mode `replace` ([line](https://github.com/xkbcommon/libxkbcommon/blob/069777f47fbb0393f88e114f101be60504dd367c/src/xkbcomp/symbols.c#L368))
* `a`'s mode in all other cases

Is this intentional? Should this be documented somewhere?

(*) I hope these examples makes sense at all, since I'm still trying to understand the format

**Comment by bluetech** [comment]
Your analysis seems correct to me.

> Is this intentional?

This has been decided 25 years ago so it's hard to know for sure. But the behavior does make sense to me.

> Should this be documented somewhere?

Currently it isn't documented in this repo. It would be nice to include details on merging in this document https://github.com/xkbcommon/libxkbcommon/blob/master/doc/keymap-format-text-v1.md but it would require some work to untangle it from the code :)

**Comment by wismill** [comment]
So I took a deep dive. In the following, I kept the naming convention from the code: `into` is the old (previous) entry and `from` is the new entry.

The rules are:
- Every `KeyInfo` starts with merge mode `default`, unless set explicitly, i.e. `<mode> key …`.
- When merging, `default` is equivalent to `override`.
- When merging two keys, the resulting merge mode is unchanged, unless the key `from` has merge type `replace`.
- When merging two symbols files, the `from` *keys* merge mode is left unchanged, *unless* the *files* merge mode is not `default`. E.g. in `include "a+b"` and `override "b"`, keys in `a` have their merge mode unaffected but keys in `b` will be change to merge mode `override`. This is important, because it mean any `replace` or `augment` in `b` will be thus ineffective. This gave me headaches until I understood it.

> What about other mode combinations? For example, what if foo's mode was augment instead of replace? And so on.

It will work as intended, unless the map is not the first one (see above). Examples:

```
// a & b unchanged, merged to parent with merge mode `default`
// Every key in the parent keeps its merge mode, unless an included key has merge mode `replace`.
include "a"
include "b"

// a: unchanged, b: override, result is merged with mode `default`
// Every key in the parent keeps its merge mode, unless an included key (in `a` only) has merge mode `replace`.
include "a+b" 
```

```
// a & b merged to parent using merge mode `augment`
// All their keys have their merge mode changed to `augment`.
// Every key in the parent keeps its merge mode.
augment "a"
augment "b"

// a: unchanged, b: override, result is merged with mode `augment`
// Every key in the parent keeps its merge mode.
augment "a+b" 
```

Etc.

---

I am working on a test suite to ensure this. Note that we should make sure `xkbcomp` passes the test suite too, unless we found an unsound behavior.

**Comment by wismill** [comment]
Personally I was very surprise that e.g. a `replace key …` in `a` is ineffective in `include "x+a"`. Well, it is not *totally* ineffective, but it cannot *reset* a key, only override it.

I wonder if we should change the meaning of `+` to be merge mode `default` instead of `overriding`. This would make all the `replace` effective in xkeyboard-config, hopefully fixing issues and not raising new ones.

**Comment by mahkoh** [comment]
>When merging two symbols files, the from keys merge mode is left unchanged, unless the files merge mode is not default. E.g. in include "a+b" and override "b", keys in a have their merge mode unaffected but keys in b will be changed to merge mode override. 

Not quite. If `a` contains no keys, then `a+b` is the same as `b`, i.e. the merge mode is unchanged.

--------------

If you are looking into fixing issues with merge modes, then you might want to look into this case which exposes a xkbcommon-only bug:

```
key <a> { };
key <a> { };
augment key <a> { [A] };
```

---------------

FYI I'm currently working on a rust implementation of XKB and I'm deliberately breaking with xkbcommon when it comes to merging behavior. I've done a deep-dive analysis of the merge behaviors in xkbcommon and my conclusion was that most of it was either ineffective (as you already found with the "a+b") or bugged in some ways.

My rules for merging are:

1. There are three merge modes: override, agument, and replace, with the other modes being aliases for override.
2. Merge modes only affect the current file, they are never propagated upwards in the include tree.

<details>
<summary>Notes regarding the xkbcommon merge behavior</summary>

```
vmod:
    local:
        merge = (merge == MERGE_DEFAULT ? stmt->merge : merge);
        merge = MERGE_OVERRIDE except in xkb_types (probably bug)
        merge mode is not persisted

    in include:
        no hierarchy. everything directly into the top-level object

keycode:
    local:
        merge = stmt->merge || MERGE_OVERRIDE;
        merge mode is not persisted

    in include:
        if duplicate key code:
            if AUGMENT:
                return
            remove old name/code
        if duplicate key name:
            if !OVERRIDE:
                return (even for REPLACE!!)
            replace key code

key alias:
    local:
        merge = MERGE_OVERRIDE;
        merge is persisted

    in include:
        if current alias list is empty:
            merge = persisted_merge
        else:
            merge = include_merge || persisted_merge
        merge is persisted

key type:
    local:
        merge = stmt->merge || include_merge || MERGE_OVERRIDE
        merge is persisted

    in include:
        if current key type list is empty:
            merge = persisted_merge
        else:
            merge = include_merge || persisted_merge
        merge is persisted
        

symbol:
    local:
        merge = stmt->merge
        if conflict and merge != MERGE_REPLACE:
            old merge is retained
        else:
            merge is persisted

    in include:
        if current symbol list is empty:
            merge = persisted_merge
        else:
            merge = include_merge || persisted_merge
        if conflict and merge != MERGE_REPLACE:
            old merge is retained
        else:
            merge is persisted

mod map def:
    local:
        merge = stmt->merge
        (DEFAULT acts like OVERRIDE)
        if conflict:
            old merge is retained
        else:
            merge is persisted

    in include:
        if current list is empty:
            merge = persisted_merge
        else:
            merge = include_merge || persisted_merge
        if conflict:
            old merge is retained
        else:
            merge is persisted

interp def:
    local:
        merge = stmt->merge || MERGE_OVERRIDE
        if conflict and merge != MERGE_REPLACE:
            old merge is retained
        else:
            merge is persisted

    in include:
        if current list is empty:
            merge = persisted_merge
        else:
            merge = include_merge || persisted_merge
        if conflict and merge != MERGE_REPLACE:
            old merge is retained
        else:
            merge is persisted

led map def:
    local:
        merge = stmt->merge || MERGE_OVERRIDE
        if conflict and merge != MERGE_REPLACE:
            old merge is retained
        else:
            merge is persisted


    in include:
        if current list is empty:
            merge = persisted_merge
        else:
            merge = include_merge || persisted_merge
        if conflict and merge != MERGE_REPLACE:
            old merge is retained
        else:
            merge is persisted
```

</details>

**Comment by wismill** [comment]
> FYI I'm currently working on a rust implementation of XKB

@mahkoh Interesting! Is there any public repository? Why not proposing the project in this organization? 

As for your other comments, I am working on an exhaustive test suite for merge behaviors and will surely add your use cases, thanks.

I did broke the merge behavior while adding support for multiple actions by level, so while I work on a fix please only compare xkbcommon before the offending commit or just use the 1.7 release.

**Comment by mahkoh** [comment]
>Is there any public repository?

Not quite yet. While the compiler (equivalent to xkbcomp) is finished, I'm still working on tests and implementing the state machine.

>Why not proposing the project in this organization?

That might be possible at some point in the future when I've stabilized the design.

**Comment by mahkoh** [comment]
>Else entries such as replace key... are useles. Well they do not work fully currently, so let's fix that.

They have never worked for anyone using RMLVO. I believe it follows that almost nobody relies on them. Therefore simplifying the rules seems worthwhile.

I don't believe anyone actually understands how the rules currently work.

**Comment by wismill** [comment]
If you do this, I do not understand why you would even bother to handle any merge mode other than the syntax. 

They are bug reports related to this, so it is likely that some were fixed by tweaking the keymaps rather than fixing xkbcomp.

**Comment by mahkoh** [comment]
I have nothing against merge modes in general. So one might reasonably write `replace "my-indicator"` to completely replace an indicator that is part of the `complete` include.

What I'm not going to do is implement merge modes that travel upwards in the include tree because they make local reasoning impossible and they have been implemented incorrectly for decades which makes it impossible to know what behavior anyone actually relies on.

**Comment by wismill** [comment]
**Local merge modes _are_ useless**. I am quite sure that the xkb creators have not bothered to implement local merge modes to be useless. I will try to find the relevant doc.

Anyway, since the keymap format in Wayland is the one supported in xkbcommon, once fixed here your implementation will be non-conformant.

**Comment by mahkoh** [comment]
>Local merge modes are useless.

They are not useless. Overwriting an indicator and replacing an indicator, for example, behave very differently and lead to different outcomes.

>Anyway, since the keymap format in Wayland is the one supported in xkbcommon

Wayland only deals with fully resolved maps. Merge modes are irrelevant at that stage.

**Comment by wismill** [comment]
Please explain me

> They are not useless. Overwriting an indicator and replacing an indicator, for example, behave very differently and lead to different outcomes.

Thank you for the lesson on the basics. I am only the main active maintainer 😅.

Could you go further and explain me how being able to use (in your implementation) e.g. `replace key` _**only**_ in the **same** `file(map)` is useful? Why would someone replace something they wrote just before? Or did I miss something? Could you provide a *real* use case on e.g. the indicators you took as example?

**Comment by mahkoh** [comment]
>Could you go further and explain me how being able to use (in your implementation) e.g. replace key only in the same file(map) is useful?

For example

```
xkb_compat {
    include "complete"
    replace indicator "Caps Lock" {
        groups = Group2;
   };
};
```

Without the `replace` this would add the group condition to the usual mods condition from the include. But the user only won't the indicator to be active if the second group is active.

**Comment by wismill** [comment]
Indeed, this is a valid use case. I wish you had mentioned it earlier in the discussion.

This pattern is only present in the vendor-specific layouts in xkeyboard-config though, while the options using merge mode rely on the propagation, which is thus the main use case for in modern.

I have not done an a deep search on the various bugs, but I know at least [this one](https://gitlab.freedesktop.org/xorg/app/xkbcomp/-/issues/12) in xkbcomp.

I need to sleep on it.

Thank you for bringing the various issues to my attention and congrats for your deep dive in the xkb land. I hope that you will inform us about the evolution of your keymap compiler.  

---

> Wayland only deals with fully resolved maps. Merge modes are irrelevant at that stage.

True. I mixed the projects: what I mean is while your compositor may compile keymaps from whatever source and provide it to the clients, the standard database is xkeyboard-config, which is tested only for xkbcomp and xkbcommon. I do not know of other keymap compiler that is not niche, so xkeyboard-config is de-facto tailored for these tools, as Wayland for xkbcommon. So if you use xkeyboard-config, you may end up with a different keymap than with xkbcommon.

**Comment by mahkoh** [comment]
From the bug report:

>And using ctrl(ctrl_ac) option, "pc+us+jp:2+ctrl(ctrl_ac), produces following map.

I could not reproduce this exact example but let's instead talk about

```
        include "pc:2+ctrl(swapcaps)"
```

which produces something very similar to what the user is reporting:

```
    key <LCTL> {
        symbols[Group1]= [       Caps_Lock ],
        symbols[Group2]= [       Control_L ]
    };
```

>This is unwanted behavior.  And I think these optional setting for symbols
should use 'replace' merge mode instead of 'override' merge mode.  But in a XKB
keymap, we cannot use "replace file(section)" notation.  Only "+ file(section)"
is available, which means 'override' merge mode.

This is incorrect. The user can already use

```
        include "pc:2"
        replace "ctrl(swapcaps)"
```

which produces the desired output

```
    key <LCTL> {         [       Caps_Lock ] };
```

What is indeed not possible is using the shorthand `a+b` to achieve this. But the situations where this shorthand is useful are even more rare than persisted merge modes in general. Indeed, `include "a"` followed by `override "b"` is only different from `include "a+b"` if the includes are preceded by other symbol decls which is never the case with RMLVO.

Adding inline notation for this, e.g. `include "a!b"`, would be possible but not backwards compatible.

**Comment by mahkoh** [comment]
Addendum: It is unclear if the user even wants this replace behavior. Would he not also want LTCL to map to Caps_Lock in the second group? Maybe what he wants is

```
        include "pc:2"
        replace "ctrl(swapcaps)+ctrl(swapcaps):2"
```

which produces, with libxkbcommon, the desired

```
	key <LCTL>               {
		symbols[Group1]= [                      Caps_Lock ],
		symbols[Group2]= [                      Caps_Lock ],
	};
```

In xkbcomp, this seems to be affected by a different bug. It produces

```
    key <LCTL> {         [       Caps_Lock ] };
```
